src/hotspot/share/oops/inlineKlass.cpp

```diff
@@ -423,19 +422,14 @@ int InlineKlass::collect_fields(GrowableArray<SigEntry>* sig, float& max_offset,
       SigEntry::add_entry(sig, T_METADATA, name(), base_off);
       SigEntry::add_entry(sig, T_VOID, name(), offset);
     }
-    max_offset = MAX2(max_offset, (float)offset);
   }
   int offset = base_off + size_helper()*HeapWordSize - (base_off > 0 ? payload_offset() : 0);
   // Null markers are no real fields, add them manually at the end (C2 relies on this) of the flat fields
   if (null_marker_offset != -1) {
-    max_offset += 0.1f; // We add the markers "in-between" because they are no real fields
-    SigEntry::add_entry(sig, T_BOOLEAN, name(), null_marker_offset, max_offset);
+    SigEntry::add_null_marker(sig, name(), null_marker_offset);
     count++;
   }
   SigEntry::add_entry(sig, T_VOID, name(), offset);
-  if (base_off == 0) {
```

Comment:
Nice, I have a pending issue of making this sort less a minefield but removing it completely is absolutely better. The comment of this function says it sorts the flattened fields in the increasing offset order, please change it, too.